### PR TITLE
Plane: On Critical Battery, Land

### PR DESCRIPTION
**Pull request by VelardeR**
Created a critical voltage and critical mAh parameter. When these are crossed, the plane enters landing sequence if present, otherwise RTL. See #3950.

### CODE REVIEW

**Comment by WickedShell** [comment]
This is using a separate set of FS_BATT_CRT_* parameters rather then adding a RTL_AUTO_LAND 3 option. I think this is the better way to go as I would prefer a vehicle RTL well before the critical battery level and allow me time to handle it as a user rather then only executing the RTL on a critical level of battery. (Basically this approach to implementation provides a second stage of battery failsafe).

**Comment by magicrub** [comment]
related: https://github.com/ArduPilot/ardupilot/issues/3561

File: libraries/AP_BattMonitor/AP_BattMonitor.h [hunk 1]
```diff
@@ -98,9 +106,14 @@ class AP_BattMonitor
...
-    bool exhausted(uint8_t instance, float low_voltage, float min_capacity_mah);
-    bool exhausted(float low_voltage, float min_capacity_mah) { return exhausted(AP_BATT_PRIMARY_INSTANCE, low_voltage, min_capacity_mah); }
+    /// status - returns BattMonitor_STATUS_CRITICAL if the voltage remains below the critical_voltage
+    ///                  for 10 seconds or remaining capacity falls below critical_capacity_mah.
+    ///          returns BattMonitor_STATUS_LOW if the voltage remains below the low_voltage for 10 
+    ///                  seconds or remaining capacity falls below min_capacity_mah.
+    ///          returns BattMonitor_STATUS_NORMAL otherwise.
+    enum BattMonitor_Status status(uint8_t instance, float low_voltage, float critical_voltage, float min_capacity_mah, float critical_capacity_mah);
+    enum BattMonitor_Status status(float low_voltage, float critical_voltage, float min_capacity_mah, float critical_capacity_mah) { 
+        return status(AP_BATT_PRIMARY_INSTANCE, low_voltage, critical_voltage, min_capacity_mah, critical_capacity_mah); }
```
Comment (magicrub) [inline libraries/AP_BattMonitor/AP_BattMonitor.h]: when doing multi-line functions in header, move the last } below it.

File: libraries/AP_BattMonitor/AP_BattMonitor.cpp [hunk 2]
```diff
@@ -293,20 +312,20 @@ bool AP_BattMonitor::exhausted(uint8_t instance, float low_voltage, float min_ca
         if (state[instance].low_voltage_start_ms == 0) {
             state[instance].low_voltage_start_ms = AP_HAL::millis();
         } else if (AP_HAL::millis() - state[instance].low_voltage_start_ms > AP_BATT_LOW_VOLT_TIMEOUT_MS) {
-            return true;
+            return BattMonitor_STATUS_LOW;
         }
     } else {
         // acceptable voltage so reset timer
         state[instance].low_voltage_start_ms = 0;
+        state[instance].critical_voltage_start_ms = 0;
```
Comment (magicrub) [inline libraries/AP_BattMonitor/AP_BattMonitor.cpp]: This also needs to be reset on it's own for proper debounce

**Comment by magicrub** [comment]
I really like this PR. I'm thinking the param should go into battery library. Low battery should go there too. That would make it common for copter too. We don't want a FS_BATT* but that might be better done on another PR.

What I'd really like to see is an auto calculation of mAh per distance and then a failsafe being called for low/critical battery that depends on distance away from home/Land point.

File: ArduPlane/events.cpp [hunk 3]
```diff
@@ -146,13 +146,38 @@ void Plane::low_battery_event(void)
...
+void Plane::critical_battery_event(void)
+{
+    if (failsafe.critical_battery) {
+        return;
+    }
+    gcs_send_text_fmt(MAV_SEVERITY_WARNING, "Critical battery %.2fV used %.0f mAh",
+                      (double)battery.voltage(), (double)battery.current_total_mah());
+    if (flight_stage != AP_SpdHgtControl::FLIGHT_LAND_FINAL &&
+        flight_stage != AP_SpdHgtControl::FLIGHT_LAND_PREFLARE &&
+        flight_stage != AP_SpdHgtControl::FLIGHT_LAND_APPROACH) {
```
Comment (magicrub) [inline ArduPlane/events.cpp]: also needs to check for ::FLIGHT_LAND_ABORT. If you're ever in that state then you're already at home trying to land

Comment (WickedShell) [inline ArduPlane/events.cpp]: @magicrub Should that thought be applied to the low_battery_event() as well? (Line 146 of this same file) I can see the arguments either way on that one...

File: ArduPlane/events.cpp [hunk 4]
```diff
@@ -146,13 +146,38 @@ void Plane::low_battery_event(void)
...
+        set_mode(RTL, MODE_REASON_BATTERY_FAILSAFE);
+        aparm.throttle_cruise.load();
+
+        // Go directly to the landing sequence
+        jump_to_landing_sequence();
```
Comment (magicrub) [inline ArduPlane/events.cpp]: remove jump_to_landing_seq and auto_state.checked_for_auto. They are handled when RTL is set and RTL_AUTOLAND >0.

Comment (WickedShell) [inline ArduPlane/events.cpp]: @magicrub this is meant to be independent of RTL_AUTOLAND

**Comment by magicrub** [comment]
@tridge and @WickedShell how do you feel about moving the low/crit battery level into AP_BATTERY?

**Comment by WickedShell** [comment]
@magicrub Would you move the FS_BATT_ params there as well? An advantage I could see in consolidating them into the BattMonitor is that you could have a low/critical event for each battery monitor which would actually be better for VTOL stuff/anything where you really *need* the second battery.

File: ArduPlane/events.cpp (hunk 4 above)
Comment (magicrub) [inline ArduPlane/events.cpp]: we don't want it to autoland if RTL_AUTOLAND == 0, it should come back home and loiter like a simple RTL.

**Comment by magicrub** [comment]
yeah, independent low/crit levels for each battery make sense.

**Comment by WickedShell** [comment]
It would mean losing the name of FS_BATT_VOLTAGE/FS_BATT_MAH though, even with the translation table, which is the biggest concern I had with that.

File: ArduPlane/events.cpp (hunk 4 above)
Comment (WickedShell) [inline ArduPlane/events.cpp]: RTL_AUTOLAND 3 being this new behaviour would be fine. The objection is that none of the current RTL_AUTOLAND options allow RTL without also doing a landing. I want RTL on low battery, but I want to have time before just going into the landing sequence, and on normal RTL there should be a option that says loiter indefinitely at a rallypoint/home until battery is low, then land. That is primarily what this is meant to address. If you think RTL_AUTOLAND 3 meaning perform a normal RTL until critical battery and only then perform a autoland I'm fine with that.

**Comment by magicrub** [comment]
If we move FS low/crit voltage to battery library then the GCSs will need to be aware of this change.

**Comment by magicrub** [comment]
This was discussed during the weekly dev call. Your PR is touching on a subject that we've been wanting to do for a while and this PR addresses
- We'd like to see the failsafe voltage moved to AP_BATTERY
- AP_Battery needs an enum for battery status (which you've done)
- AP_Battery needs better smart battery support (extends the status enum)
- Failsafe events are in platform (currently done)
- Failsafe triggers for battery need to read battery status events (TODO)

So, this PR does a little bit of that but we need to see what we can pull in and what we'd like seen changed to better suit our long-term goal of this library

**Comment by magicrub** [comment]
btw, I did not mean to sound like I/we wanted those things changed in this PR, only that they should be in mind. We need to figure out how to tweak in this PR to not inhibit that future

**Comment by amilcarlucas** [comment]
This needs a rebase. But my gues is that most of this changes have wandered to other PRs. @WickedShell can this be closed ?

**Comment by magicrub** [comment]
This is a case of someone throwing code at us and walking away.

**Comment by WickedShell** [comment]
#7213 will be how this is moving forwards.

**Pull request closed without merging**